Add number of errors to ignore while choosing replicas

azat opened the pull request:
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add number of errors to ignore while choosing replicas (`distributed_replica_error_ignore`)

Detailed description / Documentation draft:
This will allow avoid switching to another replica in case of error
(since error can be temporary).

Refs: #10564

<details>

Previous HEAD's:
- 09f8f7227297989aaed45d8fc943eafa01bab779
- c851407ffe0624f54567c71666a74874a8de8acf

</details>

azat committed:
Fix divizion by zero for distributed_replica_error_half_life=0

azat committed:
Add number of errors to ignore while choosing replicas (distributed_replica_max_ignored_errors)

This will allow avoid switching to another replica in case of error
(since error can be temporary).

azat committed:
Add a test for distributed_replica_max_ignored_errors

azat committed:
Drop unused PoolWithFailoverBase<>::reportError()

akuzm commented:
> FWIW right now there `distributed_replica_error_cap` that is the high limit for the error numbers, so **how about rename these settings** to (not sure what is the policy for settings renames):
> 
> * distributed_replica_error_ignore -> distributed_replica_error_limit_low
> * distributed_replica_error_cap    -> distributed_replica_error_limit_high

Renaming settings is almost never good -- we strive to be backwards compatible where possible.

Speaking about the names themselves, "cap" is a saturation limit for the number of errors we track for a replica. It's limited so that the error number decays in a sane time, and the replica is not considered broken forever. I think it's not complementary to your new setting, and they don't form any meaningful range, so giving them complementary names would be misleading. I'd name your new setting "distributed_replica_max_ignored_errors". Although it's kind of hard to tell from these names which limit is which...

akuzm commented on src/Common/PoolWithFailoverBase.h:
```diff
@@ -111,17 +111,16 @@ class PoolWithFailoverBase : private boost::noncopyable
-
-    void reportError(const Entry & entry);
+            bool fallback_to_stale_replicas = true,
+            size_t error_ignore = 0);
```
Do you mind making them explicit? We're constantly finding forgotten default parameters all over the code. IIRC we even had this exact problem when we added "distributed_errors_cap" to the constructor of this class.

azat replied:
>Do you mind making them explicit? We're constantly finding forgotten default parameters all over the code.

Great idea, initially it was added just like `fallback_to_stale_replicas` (with default value)

azat replied:
Done (plus some other cleanup)

P.S. I'm trying to keep PRs without cleanups/refactoring to reduce number of changes (lots of changes is hard to review), but looks like it is not a problem for you, so I will consider deviating from this rule in some cases

akuzm commented on src/Client/ConnectionPoolWithFailover.cpp:
```diff
@@ -208,7 +208,10 @@ std::vector<ConnectionPoolWithFailover::TryResult> ConnectionPoolWithFailover::g
     bool fallback_to_stale_replicas = settings ? bool(settings->fallback_to_stale_replicas_for_distributed_queries) : true;
 
-    return Base::getMany(min_entries, max_entries, max_tries, try_get_entry, get_priority, fallback_to_stale_replicas);
+    return Base::getMany(min_entries, max_entries, max_tries,
```
By the way, why is it not a property of the pool, like the other error-related parameters?

azat replied:
I thought about this, but in this case you cannot change it for specific query (it is not that relevant for cap, since it this is about restoring from errors, so it should not affect the current query)

azat commented:
>Renaming settings is almost never good -- we strive to be backwards compatible where possible.

Will remember.

>Speaking about the names themselves, "cap" is a saturation limit for the number of errors we track for a replica. It's limited so that the error number decays in a sane time, and the replica is not considered broken forever. I think it's not complementary to your new setting, and they don't form any meaningful range, so giving them complementary names would be misleading.

Yeah, when I was writing this I have some similar thoughts, ok - was a bad idea anyway.

>I'd name your new setting "distributed_replica_max_ignored_errors". Although it's kind of hard to tell from these names which limit is which...

Ok, will rename

azat committed:
Drop default values for some arguments of PoolWithFailoverBase::getMany

azat committed:
Mark PoolWithFailoverBase::get() protected

azat committed:
Pass max_ignored_errors/fallback_to_stale_replicas to PoolWithFailoverBase::get() too

azat committed:
Fix test_distributed_load_balancing flaps (due to config reload)

At startup, server loads configuration files.

However ConfigReloader does not know about already loaded files (files
is empty()), hence it will always reload the configuration just after
server starts (+ 2 seconds, reload timeout).

And on configuration reload the clusters will be re-created, so some
internal stuff will be reseted:
- error_count
- last_used (round_robing)

And if the reload will happen during round_robin test it will start
querying from the beginning, so let's issue config reload just after
start to avoid reload in the middle of the test execution.

azat commented:
>ClickHouse build check — 11/17 builds are OK

Some CI issues
The only difference between b8ee2ea (everything is green) and bd45592 (this build fail) is the integration test anyway

alexey-milovidov commented:
I'm not sure how admin is supposed to choose the value of this setting.

nvartolomei commented:
I think a better way to balance things would be to track the rate of errors (instead of an absolute value) with a sliding window and allow a small variability in the rate of errors between replicas.

The ignore value added here seems to help only in the beginning before the err counter reaches that value and then the behaviour will be exactly the same as it currently is until the counters come back to a values less than this min constant.

alexey-milovidov commented:
@nvartolomei The calculated number of errors is already a sliding "window" with exponential smoothing.

azat commented:
>I'm not sure how admin is supposed to choose the value of this setting.

via `config.xml`?

alexey-milovidov commented:
What value will be good: 1 or 10 or 100?

azat commented:
It depends, and should be adjusted only after looking at system.clusters (that has error_count rate) + server logs (for issues around distributed queries).

I would say that 100 definitely is not good, but 1-10 may be a good idea.

Also rate of queries and time of replica error_count recover (distributed_replica_error_half_life) also should be take into account.

(at first glance this can be enough)

akuzm committed:
Merge remote-tracking branch 'origin/master' into HEAD

akuzm committed:
remove unneeded define

akuzm commented:
> P.S. I'm trying to keep PRs without cleanups/refactoring to reduce number of changes (lots of changes is hard to review), but looks like it is not a problem for you, so I will consider deviating from this rule in some cases

That's generally a good idea if there is a non-trivial amount of refactoring. One way to do it is maintain logically separate changes in separate commits -- it works well with the github interface. For very big PRs, we can even merge some commits earlier, to have less conflicts. It's not very convenient for the developer though -- lots of fiddling with interactive rebase.

azat commented:
>That's generally a good idea if there is a non-trivial amount of refactoring.

And sometimes even trivial may be a problem (especially for backporting)

> One way to do it is maintain logically separate changes in separate commits -- it works well with the github interface.

Well, keeping separate changes in separate commit is a rule of thumb, but this also does not solves issues with conflicts and so on.

> lots of fiddling with interactive rebase.

ccache helps at least here :)
